### DISCUSSION TITLE
client:visible + no server side rendering option.

### DISCUSSION THREAD

**Discussion by angelcervera** [withastro/roadmap#501] [Proposal]
# Summary

Using `client:only`, we avoid running on the server side, but on the client side is loaded immediately. It would be great to have the same option for `client:visible` to avoid to run in client side.

# Background & Motivation

I have a component that reads from the database. This type of calls, or other heavy APIs calls are slow, heavy, and expensive (in comparison to CDN calls) calls, so I want to load and run and load that Island only **when it is visible but avoid running it in the server-side**. Remember that is, as an example, a call to a Database that will not generate static data.

I used `client:visible` expecting that behavior, but it runs on the server side as well. Using `client:only="vue"` skip running in the server, but it loads immediately after loading the page.

After asking in Discord, looks like this option is not implemented. :(

# Goals

- Add new client directive or upgrade the behavior of `client:visible` but with the option of skipping server-side render, like `client:only` is doing.

**Comment by preetamslot** [comment]
This request also covers this:
https://github.com/withastro/roadmap/discussions/272

**Comment by carvinlo** [comment]
Hello, any updates on this?
Or allows users to use `addClientOnlyDirective` that are rendered only on the client.

**Comment by rapidfixer** [comment]
According to the source code, there is a hardcoded check for this exact directive to disable server rendering:
https://github.com/withastro/astro/blob/2a0fa86bacfab1d9c66d20ebf5920ab629478e06/packages/astro/src/runtime/server/jsx.ts#L139
As a simple solution, i would suggest to change it to prefix check, so then user can create own directive using existing API something like: `client:only:oninteraction`.

**Reply by delucis** [reply]
I think an attribute value might work well here too as an option:

```astro
<Component client:only={{ on: 'interaction' }} />
```

Although I’ll note there are some tricky nuances. How do you have “on interaction” for a component that hasn’t rendered yet? And similarly, when is a component “visible” if it hasn’t rendered yet?
